fixes #3338 - grey out VM form when editing is not supported

lzap opened the pull request:
I went for JavaScript solution because it allows us a generic solution.
ComputeResources already have supports_update? methods and it is easy to
determine this in one place rather than editing all the fields and all
possible places. This will prevent errors when implementing new CRs.

skottler commented:
Just tested this out and it looks great. Thanks @lzap!

What do you think about adding some kind of notice at the top of the view explaining why the fields are locked?

ohadlevy commented on app/views/hosts/_compute.html.erb:
```diff
@@ -14,4 +14,8 @@
   <%= hidden_field_tag 'capabilities', compute_resource.capabilities %>
+
+  <% if @host && !compute_resource.supports_update? %>
+    <script type="text/javascript">$(document).ready(disable_vm_form_fields)</script>
```
I would prefer to make it unintrusive, maybe using a wrapper div and a data-attribute.

e.g.

``` ruby
content_tag :div, :data => {:supports_update => @compute_resource.supports_update?} do
   # fields go here
end
```

and then on your javascript

``` javascript
$(.form).find('[data-supports-update=false]').attr('disabled', true)
```

not tested at all, and open for feedback.. @abenari ?

lzap replied:
But this approach completely flips around purpose of my thinking - I did
not want to put anything into specific CR edit forms. This way it works
"automatic", even when we add a new CR form there.

I understand that this is not best approach for everything, cos it is
importing some "magic", but in this case it is not critical feature.
Just a cosmetic thing.

Disabled elements have title set, when you hover to them it appears
usually. But a notice is a good point. I will what what @abenari says.

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

ohadlevy replied:
not sure i follow? this would be in the same place? just making the js more generic instead?

lzap replied:
Yeah, since this is one-liner currently I agree.

But Dominic points out also on the +Add Interface and similar buttons,
then the function can get more complicated. I'd rather keep it separate.

Please confirm if you like this pure-javascript approach and I will go
ahead and remove all the static things in vmware form and add button
handling.

LZ

On Mon, Oct 21, 2013 at 01:01:57AM -0700, Ohad Levy wrote:

> > @@ -14,4 +14,8 @@
> >      </div>
> >    <% end%>
> >    <%= hidden_field_tag 'capabilities', compute_resource.capabilities %>
> > +
> > -  <% if @host && !compute_resource.supports_update? %>
> > -    <script type="text/javascript">$(document).ready(disable_vm_form_fields)</script>
> 
> not sure i follow? this would be in the same place? just making the js more generic instead?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/theforeman/foreman/pull/959/files#r7087496

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

ohadlevy replied:
I dont have a strong opinion here, @abenari any comment?

domcleal commented:
Something similar was done in #901 for VMware specifically, but I like that this is becoming generic.  It'd be good if we could revert the additional "disabled" attributes added by that PR to the VMware form in this one.

One thing it doesn't prevent which #901 does, is the addition and removal of network interfaces and storage volumes.  Can we also disable these links?

abenari committed:
fixes #3338 - grey out VM form when editing is not supported

abenari commented:
1. I would prefer to use unobtrusive approach.
2. I don't like the title on every input that explains why it is disabled, I prefer one alert message at the top.
3. Better remove the add/remove network and disk buttons then making them disabled.

Here is my suggestion:
https://github.com/abenari/foreman/commit/824ec9b59fa738ba0e9e7df3954e9a3a7b77258b

lzap commented:
I pushed your solution, looks better for sure. Works well.

I am fuzzy about how to disable "+ Add Volume" buttons and others. I was
thinking about moving add_child_node into a function and doing
something:

  if ($("#update_not_supported") &&
    $("#update_not_supported").is(":visible")) {
    // add_child_node...
  }

This is different for each CR, but the approach seems to be similar.
Inline of via function call to add_child_node.

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

ohadlevy commented:
@lzap, @abenari where does this stands? thanks

lzap commented:
Need advice from @abenari about how to disable buttons. The bit I sent above does not seem to be right.

ohadlevy commented:
@abenari ping ?

abenari commented:
@lzap, to disable the buttons you can use this code:
if ($('#supports_update').data('supports-update')){
// add child node 
}
However, I prefer hiding the buttons rather than disabling them.
Disabled buttons suggest there is something the user can do to enable them.
this line https://github.com/abenari/foreman/commit/824ec9b59fa738ba0e9e7df3954e9a3a7b77258b#diff-63d9ef3bbe6ac020529a197ac7aa171bR621 should be removing the buttons if it doesn't we should fix it.

lzap commented:
Ok I changed the behavior of two things:

A) The way we hide X and + buttons. Now we find the relevant A tags via named archor "disable_vm". If we want to enable hiding for any particular A element, we can add that. I added for libvirt and vmware.

B) Your patch did not work in case when there was error in the new host form - it was disabling the VM form, so I added new_record? clause to the data attribute. Now it works fine.

I think this solution is the final, hopefully :-D

domcleal commented:
Ok, this is looking good.  Could you please partially revert 29769a4 too, so we don't have two implementations?  Specifically any disabled attributes or checks for "new".

A minor thing, but this could simply use a class rather than the value of the href attribute, so:

```
 :class => "info disable_vm"
```

and then:

```
$("a[.disable_vm]").remove();
```

lzap commented:
Done, can you test this against VMWARE? I dont have any instance.

LZ

On Fri, Nov 01, 2013 at 02:27:20AM -0700, Dominic Cleal wrote:

> Ok, this is looking good.  Could you please partially revert 29769a4 too, so we don't have two implementations?  Specifically any disabled attributes or checks for "new".
> 
> A minor thing, but this could simply use a class rather than the value of the href attribute, so:
> 
> ```
>  :class => "info disable_vm"
> ```
> 
> and then:
> 
> ```
> $("a[.disable_vm]").remove();
> ```
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/theforeman/foreman/pull/959#issuecomment-27554894

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

domcleal commented on app/views/compute_resources_vms/form/_libvirt.html.erb:
```diff
@@ -12,7 +12,7 @@
       <%= render 'compute_resources_vms/form/libvirt/network', :f => i, :compute_resource => compute_resource %>
     <% end %>
-    <%= add_child_link '+ ' + _("Add Interface"), :nics, { :class => "info", :title => _('add new network interface') } %>
+    <%= add_child_link '+ ' + _("Add Interface"), :nics, { :class => "info disable-when-new", :title => _('add new network interface'), :href => '#' } %>
```
Is the href=# now obsolete?

domcleal commented on app/assets/javascripts/host_edit.js:
```diff
@@ -612,3 +613,11 @@ function interface_type_selected(element) {
+  $("[id^=host_compute_attributes]").each(function () {
+    $(this).attr("disabled", "disabled");
+  });
+  $("a[.disable-when-new]").remove();
```
Sorry, you shouldn't trust my JS advice :)

This should have been `a.disable-when-new`.  Tested with VMware.

lzap replied:
I am very bad javascript coder... I have to trust everyone. :-P

Thanks for the tip, this is much much nicer.

Fixed all your remarks, works like a charm, but I have NOT tested
vmware.

LZ

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

abenari commented:
merged in 41e445fe71487d5e79e062802f93a6217e654df8
thanks @lzap
